jetty-util/src/main/java/org/eclipse/jetty/util/SearchPattern.java

```diff
@@ -0,0 +1,179 @@
+     * @param data The data in which to search for. The data may be arbitrary binary data, 
+     * but the pattern will always be {@link StandardCharsets#US_ASCII} encoded.
+     * @param offset The offset within the data to start the search
+     * @param length The length of the data to search
+     * @return The index within the data array at which the first instance of the pattern or -1 if not found
+     */
+    public int match(byte[] data, int offset, int length)
+    {
+        validate(data, offset, length);
+
+        int skip = offset;
+        while(skip <= offset+length - pattern.length)
+        {            
+            for(int i = pattern.length-1; data[skip+i] == pattern[i]; i--) 
+                if(i==0) return skip;
+            
+            skip += table[data[skip + pattern.length - 1]];
+        }
+        
+        return -1;
+    }
+    
+    /**
+     * Search for a partial match of the pattern at the end of the data.
+     * @param data The data in which to search for. The data may be arbitrary binary data, 
+     * but the pattern will always be {@link StandardCharsets#US_ASCII} encoded.
+     * @param offset The offset within the data to start the search
+     * @param length The length of the data to search
+     * @return the length of the partial pattern matched and 0 for no match.
+     */
+    public int endsWith(byte[] data, int offset, int length)
+    { 
+        validate(data, offset, length);
+        
+        int skip = (pattern.length <= length) ? (offset+length-pattern.length) : offset;
+        while(skip < offset+length)
+        {       
+            for(int i = (offset+length-1)-skip; data[skip+i] == pattern[i]; --i)
+                if(i==0) return(offset+length - skip);
+            
+            if(skip + pattern.length - 1 < data.length)
+                skip += table[data[skip + pattern.length - 1]];
+            else
+                skip++;
+        }
+        
+        return 0;
+    }
+    
+    /**
+     * Search for a possibly partial match of the pattern at the start of the data.
+     * @param data The data in which to search for. The data may be arbitrary binary data, 
+     * but the pattern will always be {@link StandardCharsets#US_ASCII} encoded.
+     * @param offset The offset within the data to start the search
+     * @param length The length of the data to search
+     * @param matched The length of the partial pattern already matched
+     * @return the length of the partial pattern matched and 0 for no match.
+     */
+    public int startsWith(byte[] data, int offset, int length, int matched)
+    {
+        validate(data, offset, length);
+        
+        int matchedCount = 0;
+        
+        for(int i=0; i<pattern.length-matched && i < offset+length; i++)
+        {
+            if(data[i] == pattern[i+matched])
+                matchedCount++;
+            else
+                break;
+        }
+        
+        return matchedCount;   
```

Comment:
I thought it was just supposed to return the length of the pattern matched in that byte array. But I can change it to return the cumulative match although it might break some of my tests.

Reply:
@gregw The javadoc currently says it will be returning 0 for no match, which doesn't really fit if returning the cumulative match count. Should I change it to returning `matched + matchedCount` and just returning `matched` if there was no match.

Reply:
I think returning 0 for no match is correct.    If the boundary is `BOUNDARY` and we have previously seen `BOUND` and call startsWith with (`ARX`,5) then the previous match of `BOUND` is no longer valid and 0 is the correct return.  So we should be able to do 
```
  charactersMatchedSoFar = sp.startsWith(buffer,charactersMatchedSoFar);
```
and get a reasonable answer, which is 0 if a match is no longer possible, or the cumulative match count if it is.
So for boundary `BOUNDARY` and we have previously seen `BOUND`:
 - startsWith(`A`,5) == 6
 - startsWith(`AR`,5) == 7
 - startsWith(`ARY`,5) == 8
 - startsWith(`ARYxxxx`,5) == 8
 - startsWith(`ARxxxxxx`,5) == 0